### DISCUSSION TITLE
`objc[15959] fork error`

### DISCUSSION THREAD

**Discussion by dereckmezquita** [rexyai/RestRserve#215] [Q&A]
Hello, I'm having a very hard time tracking down what caused this error and I just wanted to share and discuss with others.

It's a little bit difficult for me to share a minimal example as I'm not really sure what's causing the error but here is what the error I get is:

```R
(base) work@Derecks-MacBook-Air server % Rscript src/index.R
- The project is out-of-sync -- use `renv::status()` for details.
{"timestamp":"2024-08-10 21:39:03.459255","level":"INFO","name":"Application","pid":92691,"msg":"","context":{"http_port":8080,"endpoints":{"POST":"/api/v1/trader_management/create_bot","HEAD":"/health","GET":"/health"}}}
-- running Rserve in this R session (pid=92691), 2 server(s) --
(This session will block until Rserve is shut down)

objc[92604]: +[NSNumber initialize] may have been in progress in another thread when fork() was called.
objc[92604]: +[NSNumber initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

Now, in my program, I am running a RestRserve + R6 classes for logic and one of the things I am including is a database connection.

I refactored my programme so that every time I access need a db connection a new conneciton is created to avoid problems as I know RestRserve has limitations on forking db connecitons. However, I still get the above error.

Interestingly only get the error sometimes. I get the error 4/10 times. When I do I get the error only one time when the REST API gets its first request, then after no more. I have not done extensive testing so I hold my tongue.

Currently I am running the programme and I am not seeing it anymore.

I'm hoping someone recognises the error and or might have suggestions on how to debug. I am running the programme using `Rscript` so I don't have access to `browser` or `traceback`.

**Comment by s-u** [comment]
This looks like you may be pre-loading packages or code that uses the `CoreFoundation` framework on macOS before spawning the server - which is not supported by Apple (for example, Quartz uses CF). You must not load any such code - you can only load it in the handler itself, but not before you start the server.

(E.g., if you want to use graphics, use the `Cairo` package from CRAN which has been compiled to be `fork`-safe).
